tests/ingress/ingress_test.go

```diff
@@ -0,0 +1,189 @@
+	err = cpCtrlClient.List(ctx, clusterList, client.MatchingLabels{capiv1alpha3.ClusterLabelName: clusterID})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	baseDomain := strings.TrimPrefix(clusterList.Items[0].Spec.ControlPlaneEndpoint.Host, "api.")
+	appEndpoint := fmt.Sprintf("%s.%s", helloWorldAppName, baseDomain)
+
+	var appTest apptest.Interface
+	{
+		runtimeScheme := runtime.NewScheme()
+		appSchemeBuilder := runtime.SchemeBuilder{
+			capiv1alpha3.AddToScheme,
+			capzv1alpha3.AddToScheme,
+			expcapiv1alpha3.AddToScheme,
+			expcapzv1alpha3.AddToScheme,
+			corev1.AddToScheme,
+		}
+		err := appSchemeBuilder.AddToScheme(runtimeScheme)
+		if err != nil {
+			t.Fatal(err)
+		}
+		appTest, err = apptest.New(apptest.Config{
+			KubeConfig: string(cpKubeConfig),
+			Logger:     logger,
+			Scheme:     runtimeScheme,
+		})
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	apps := []apptest.App{
+		{
+			AppCRNamespace:     clusterID,
+			AppOperatorVersion: "1.0.0",
+			CatalogName:        "giantswarm",
+			KubeConfig:         string(tcKubeConfig),
+			Name:               "nginx-ingress-controller-app",
+			Namespace:          "kube-system",
+			ValuesYAML:         fmt.Sprintf(IngressValues, baseDomain),
+			Version:            "1.11.0",
+			WaitForDeploy:      true,
+		},
+		{
+			AppCRNamespace:     clusterID,
+			AppOperatorVersion: "1.0.0",
+			CatalogName:        "default",
+			KubeConfig:         string(tcKubeConfig),
+			Name:               helloWorldAppName,
+			Namespace:          "default",
+			ValuesYAML:         fmt.Sprintf(HelloWorldValues, appEndpoint),
+			Version:            "0.3.0",
+			WaitForDeploy:      true,
+		},
+	}
+	err = appTest.InstallApps(ctx, apps)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	pod, err := createPodThatSendsHttpRequestToEndpoint(ctx, cpCtrlClient, clusterID, appEndpoint)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	o := func() error {
+		objectKey, err := client.ObjectKeyFromObject(pod)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		scheduledPod := &corev1.Pod{}
+		err = cpCtrlClient.Get(ctx, objectKey, scheduledPod)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		if scheduledPod.Status.Phase != corev1.PodSucceeded {
+			return microerror.Maskf(executionFailedError, "container didn't finish yet, pod state is %#q", scheduledPod.Status.Phase)
+		}
+
+		if scheduledPod.Status.ContainerStatuses[0].State.Terminated.ExitCode != 0 {
+			return microerror.Maskf(executionFailedError, "expected container exit code is 0, got %d", scheduledPod.Status.ContainerStatuses[0].LastTerminationState.Terminated.ExitCode)
+		}
+
+		return nil
+	}
+	b := backoff.NewExponential(backoff.ShortMaxWait, backoff.ShortMaxInterval)
+	n := backoff.NewNotifier(logger, ctx)
+	err = backoff.RetryNotify(o, b, n)
+	if err != nil {
+		_ = cpCtrlClient.Delete(ctx, pod)
+		_ = appTest.CleanUp(ctx, apps)
+		t.Fatalf("couldn't get successful HTTP response from hello world app: %v", err)
```

Comment:
Would it make sense to improve this in the future so that there was somewhere functionality where you could list all installed apps and those corresponding CRs (and the derived charts etc.) would be then listed or saved for debugging somewhere to make problem investigation easier in case this test fails? All that could even be behind a test flag that you specify from CLI when running test.

Reply:
We have to figure out the best way to do it, but it'd be a nice improvement. Otherwise it's hard to know what's really going on.